In [81]:
import numpy as np
import pandas as pd

In [82]:
df = pd.read_csv('Zip_Zri_MultiFamilyResidenceRental-2.csv',index_col=False)
zillow= df.copy()

In [83]:
zillow

RegionID  RegionName            City State  \
0        61639       10025        New York    NY   
1        84654       60657         Chicago    IL   
2        61637       10023        New York    NY   
3        91982       77494            Katy    TX   
4        84616       60614         Chicago    IL   
...        ...         ...             ...   ...   
1856     90755       75202          Dallas    TX   
1857     94610       84627         Ephraim    UT   
1858     84452       60301        Oak Park    IL   
1859     94629       84647  Mount Pleasant    UT   
1860     91695       77046         Houston    TX   

                                 Metro       CountyName  SizeRank  2010-09  \
0          New York-Newark-Jersey City  New York County         1   2930.0   
1             Chicago-Naperville-Elgin      Cook County         2   1447.0   
2          New York-Newark-Jersey City  New York County         3   2797.0   
3     Houston-The Woodlands-Sugar Land    Harris County         4      NaN   
4             Chicago-Naperville-Elgin      Cook County         5   1437.0   
...                                ...              ...       ...      ...   
1856       Dallas-Fort Worth-Arlington    Dallas County      1857    997.0   
1857                               NaN   Sanpete County      1858      NaN   
1858          Chicago-Naperville-Elgin      Cook County      1859      NaN   
1859                               NaN   Sanpete County      1860      NaN   
1860  Houston-The Woodlands-Sugar Land    Harris County      1861      NaN   

      2010-10  2010-11  ...  2019-04  2019-05  2019-06  2019-07  2019-08  \
0      2952.0   2926.0  ...   3484.0   3523.0   3573.0   3622.0   3664.0   
1      1465.0   1469.0  ...   1731.0   1749.0   1768.0   1787.0   1801.0   
2      2811.0   2813.0  ...   3449.0   3459.0   3479.0   3491.0   3510.0   
3         NaN      NaN  ...   1226.0   1250.0   1264.0   1251.0   1246.0   
4      1469.0   1490.0  ...   1888.0   1911.0   1934.0   1950.0   1953.0   
...       ...      ...  ...      ...      ...      ...      ...      ...   
1856   1019.0   1042.0  ...   1615.0   1619.0   1628.0      NaN      NaN   
1857      NaN      NaN  ...      NaN      NaN      NaN      NaN      NaN   
1858      NaN      NaN  ...   1447.0   1483.0   1495.0   1511.0   1537.0   
1859      NaN      NaN  ...      NaN      NaN      NaN      NaN      NaN   
1860      NaN      NaN  ...      NaN      NaN      NaN      NaN      NaN   

      2019-09  2019-10  2019-11  2019-12  2020-01  
0      3698.0   3704.0   3692.0   3715.0   3676.0  
1      1806.0   1789.0   1761.0   1747.0   1731.0  
2      3521.0   3574.0   3608.0   3561.0      NaN  
3      1252.0   1280.0   1310.0   1294.0   1286.0  
4      1952.0   1921.0   1958.0      NaN   1916.0  
...       ...      ...      ...      ...      ...  
1856   1729.0   1697.0   1739.0      NaN      NaN  
1857      NaN      NaN      NaN   1298.0      NaN  
1858   1570.0   1580.0   1625.0      NaN      NaN  
1859      NaN      NaN      NaN      NaN   1221.0  
1860      NaN      NaN      NaN      NaN      NaN  

[1861 rows x 120 columns]

In [84]:
# Huge number of rows with tons of missing values.
sorted(np.sum(zillow.isnull(), axis=1), reverse=True)

[113,
 113,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105

In [85]:
np.mean(np.sum(zillow.isnull(), axis=1) > 55)

0.23589468027941968

In [86]:
np.mean(np.sum(zillow.isnull(), axis=1) > 35)

0.5303600214938206

I did not realize the multifam data has so many missing values. half of all price rows are missing 1/3 of their values, and a quarter of all rows are missing half of their values. I really really don't know that we should use this data. Or at the very least, we should be dropping a HUGE amount of it. My interpolate function was built using the ZORI data.

In [87]:
# Well, we don't have any use for data before our first census dataset. 
# That is 2011, means sept 2012, which is a forecast for sept 2015. 
# Should delete all data older than sept 2012 (to include current rent) and then look at missingness.



zillow_trim = pd.merge(zillow.iloc[:,:7], zillow.iloc[:,28:], how='left', left_index=True, right_index=True)


In [88]:
zillow_trim.shape

(1861, 99)

In [89]:
np.mean(np.sum(zillow_trim.isnull(), axis=1) > 46)

0.1955937667920473

In [90]:
np.mean(np.sum(zillow_trim.isnull(), axis=1) > 31)

0.261687264911338

Okay, slightly more manageable. 20% of rows still have half or more values missing, and 26% of rows have 1/3 or more value missing. I think we might want to drop these 20% of rows that have 50%+ missing rent values. 

In [91]:
# Rows to drop:
zillow_trim[np.sum(zillow_trim.isnull(), axis=1) > 46]

RegionID  RegionName            City State  \
5        91940       77449            Katy    TX   
15       74101       37013       Nashville    TN   
25       92036       77573     League City    TX   
33       71067       30349       Riverdale    GA   
67       91889       77379          Spring    TX   
...        ...         ...             ...   ...   
1853     92311       78215     San Antonio    TX   
1857     94610       84627         Ephraim    UT   
1858     84452       60301        Oak Park    IL   
1859     94629       84647  Mount Pleasant    UT   
1860     91695       77046         Houston    TX   

                                           Metro        CountyName  SizeRank  \
5               Houston-The Woodlands-Sugar Land     Harris County         6   
15    Nashville-Davidson--Murfreesboro--Franklin   Davidson County        16   
25              Houston-The Woodlands-Sugar Land  Galveston County        26   
33                 Atlanta-Sandy Springs-Roswell    Clayton County        34   
67              Houston-The Woodlands-Sugar Land     Harris County        68   
...                                          ...               ...       ...   
1853                   San Antonio-New Braunfels      Bexar County      1854   
1857                                         NaN    Sanpete County      1858   
1858                    Chicago-Naperville-Elgin       Cook County      1859   
1859                                         NaN    Sanpete County      1860   
1860            Houston-The Woodlands-Sugar Land     Harris County      1861   

      2012-06  2012-07  2012-08  ...  2019-04  2019-05  2019-06  2019-07  \
5         NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
15        NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
25        NaN      NaN      NaN  ...   1157.0   1150.0   1131.0   1111.0   
33        NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
67        NaN      NaN      NaN  ...   1245.0   1238.0   1231.0   1247.0   
...       ...      ...      ...  ...      ...      ...      ...      ...   
1853      NaN      NaN      NaN  ...   1268.0      NaN   1334.0   1304.0   
1857      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
1858      NaN      NaN      NaN  ...   1447.0   1483.0   1495.0   1511.0   
1859      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
1860      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   

      2019-08  2019-09  2019-10  2019-11  2019-12  2020-01  
5         NaN      NaN   1314.0   1321.0      NaN   1353.0  
15        NaN      NaN   1028.0   1015.0   1013.0   1010.0  
25     1100.0   1092.0   1103.0   1098.0   1094.0   1110.0  
33        NaN   1120.0   1101.0   1087.0   1057.0      NaN  
67     1262.0   1296.0   1292.0      NaN      NaN      NaN  
...       ...      ...      ...      ...      ...      ...  
1853      NaN      NaN   1194.0   1188.0   1186.0      NaN  
1857      NaN      NaN      NaN      NaN   1298.0      NaN  
1858   1537.0   1570.0   1580.0   1625.0      NaN      NaN  
1859      NaN      NaN      NaN      NaN      NaN   1221.0  
1860      NaN      NaN      NaN      NaN      NaN      NaN  

[364 rows x 99 columns]

In [92]:
# Rows to keep:
zillow_trim = zillow_trim[np.sum(zillow_trim.isnull(), axis=1) <= 46]
zillow_trim

RegionID  RegionName         City State  \
0        61639       10025     New York    NY   
1        84654       60657      Chicago    IL   
2        61637       10023     New York    NY   
3        91982       77494         Katy    TX   
4        84616       60614      Chicago    IL   
...        ...         ...          ...   ...   
1850     84608       60606      Chicago    IL   
1852     62528       12307  Schenectady    NY   
1854     58656        2142    Cambridge    MA   
1855     65393       18232     Lansford    PA   
1856     90755       75202       Dallas    TX   

                                 Metro          CountyName  SizeRank  2012-06  \
0          New York-Newark-Jersey City     New York County         1   3045.0   
1             Chicago-Naperville-Elgin         Cook County         2   1526.0   
2          New York-Newark-Jersey City     New York County         3   2932.0   
3     Houston-The Woodlands-Sugar Land       Harris County         4      NaN   
4             Chicago-Naperville-Elgin         Cook County         5   1593.0   
...                                ...                 ...       ...      ...   
1850          Chicago-Naperville-Elgin         Cook County      1851      NaN   
1852           Albany-Schenectady-Troy  Schenectady County      1853      NaN   
1854           Boston-Cambridge-Newton    Middlesex County      1855      NaN   
1855        Allentown-Bethlehem-Easton       Carbon County      1856      NaN   
1856       Dallas-Fort Worth-Arlington       Dallas County      1857   1300.0   

      2012-07  2012-08  ...  2019-04  2019-05  2019-06  2019-07  2019-08  \
0      3083.0   3124.0  ...   3484.0   3523.0   3573.0   3622.0   3664.0   
1      1541.0   1562.0  ...   1731.0   1749.0   1768.0   1787.0   1801.0   
2      2958.0   2972.0  ...   3449.0   3459.0   3479.0   3491.0   3510.0   
3         NaN      NaN  ...   1226.0   1250.0   1264.0   1251.0   1246.0   
4      1607.0   1624.0  ...   1888.0   1911.0   1934.0   1950.0   1953.0   
...       ...      ...  ...      ...      ...      ...      ...      ...   
1850      NaN      NaN  ...   2145.0   2155.0   2166.0   2183.0   2206.0   
1852      NaN      NaN  ...    937.0    949.0    964.0    974.0    978.0   
1854      NaN      NaN  ...   2734.0   2720.0   2745.0   2777.0   2811.0   
1855      NaN      NaN  ...    781.0    773.0    769.0    762.0    758.0   
1856   1299.0   1319.0  ...   1615.0   1619.0   1628.0      NaN      NaN   

      2019-09  2019-10  2019-11  2019-12  2020-01  
0      3698.0   3704.0   3692.0   3715.0   3676.0  
1      1806.0   1789.0   1761.0   1747.0   1731.0  
2      3521.0   3574.0   3608.0   3561.0      NaN  
3      1252.0   1280.0   1310.0   1294.0   1286.0  
4      1952.0   1921.0   1958.0      NaN   1916.0  
...       ...      ...      ...      ...      ...  
1850   2228.0   2217.0   2230.0   2200.0   2192.0  
1852    978.0      NaN      NaN   1039.0   1047.0  
1854   2821.0   2799.0      NaN   3102.0      NaN  
1855    758.0      NaN      NaN      NaN      NaN  
1856   1729.0   1697.0   1739.0      NaN      NaN  

[1497 rows x 99 columns]

In [93]:
# Total number of NaN's in trimmed data
zillow_trim.isnull().sum().sum()

18826

In [94]:
# Fill interior NaN's
zillow_trim = pd.merge(zillow_trim.iloc[:,0:7], zillow_trim.iloc[:,8:].interpolate(limit_area='inside', axis=1), left_index=True, right_index=True)

In [95]:
# Total number of beginning / end NaN's
zillow_trim.isnull().sum().sum()

13769

In [96]:
#zillow_trim.to_csv('temp_zillow_prefill.csv')

- 2102-06 to 2020-01 time frame (need sept for each year).
- Dropped 364 of 1860 rows that had 50%+ missing values
- Initial NaN count for this time frame post-drop: 18,826 / 86,464
- Filled inside missing values using linear interpolation, missing count after: 13,769

- This ain't good. 10k more NaN's after filling middle than if I had a 2015 cutoff. Almost 1/3 of all the values from 2012-2015 are NaN's. Thats just too many. 


In [97]:
zillow_trim = zillow_trim.reset_index(drop=True)

In [98]:
zillow_trim.iloc[-50:-20, 6:]

SizeRank  2012-07  2012-08  2012-09  2012-10  2012-11  2012-12  2013-01  \
1447      1788    570.0    566.0    562.0    568.0    576.0    577.0    573.0   
1448      1791      NaN    631.0    640.0    641.0    661.8    682.6    703.4   
1449      1792    728.0    715.0    703.0    708.0    718.0    728.0    715.0   
1450      1793      NaN      NaN    725.0    711.0    699.0    680.0    664.0   
1451      1794      NaN      NaN    713.0    715.0    742.5    770.0    776.0   
1452      1795   1181.0   1188.0   1195.0   1193.0   1191.0   1181.0   1170.0   
1453      1796   1102.0   1120.0   1131.0   1112.0   1097.0   1078.0   1060.0   
1454      1797      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1455      1799   2793.0   2793.0   2780.0   2745.0   2727.0   2749.0   2815.0   
1456      1801    628.0    633.0    633.0    627.0    623.0    619.0    624.0   
1457      1802      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1458      1804      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1459      1805    868.0    867.0    875.0    883.0    883.0    876.0    874.0   
1460      1807    767.0    760.0    755.0    736.0    719.0    702.0    687.0   
1461      1808      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1462      1810      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1463      1812      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1464      1816      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1465      1818   1572.0   1572.0   1574.0   1579.0   1583.0   1584.0   1579.0   
1466      1819      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1467      1820   1513.0   1518.0   1518.0   1521.0   1529.0   1548.0   1550.0   
1468      1821      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1469      1822      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1470      1823    864.0    873.0    867.0    870.0    884.0    905.0    932.0   
1471      1825    747.0    739.0    733.0    736.0    749.0    742.0    727.0   
1472      1826      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1473      1827      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1474      1828      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1475      1829   1021.0    992.0    961.0    930.0    924.0    931.0    959.0   
1476      1830      NaN      NaN      NaN      NaN      NaN      NaN      NaN   

      2013-02  2013-03  ...      2019-04      2019-05      2019-06  \
1447   574.00    572.0  ...   869.000000   878.000000   890.000000   
1448   724.20    745.0  ...   613.000000   618.000000   627.000000   
1449   704.00    698.0  ...   791.000000   795.000000   798.000000   
1450   657.00    672.0  ...   627.000000   627.000000   632.000000   
1451   775.00    780.0  ...   940.000000   940.000000   941.000000   
1452  1157.00   1158.0  ...  1262.000000  1251.000000  1240.000000   
1453  1050.00   1042.0  ...  1211.000000  1216.000000  1225.000000   
1454      NaN      NaN  ...   881.428571   892.142857   902.857143   
1455  2882.00   2944.0  ...  3339.000000  3357.000000  3361.000000   
1456   632.00    626.0  ...   735.000000   738.000000   742.000000   
1457      NaN      NaN  ...  2057.000000  2048.000000  2047.000000   
1458      NaN      NaN  ...  1465.000000  1471.000000  1453.000000   
1459   877.00    879.0  ...  1044.000000  1061.000000  1080.000000   
1460   681.00    689.0  ...   722.000000   715.000000   714.000000   
1461      NaN      NaN  ...  2523.000000  2571.000000  2621.000000   
1462      NaN      NaN  ...  3882.000000  3906.000000  3946.000000   
1463      NaN      NaN  ...  2918.000000  2937.000000  2961.000000   
1464      NaN      NaN  ...  2529.000000  2536.000000  2541.000000   
1465  1585.00   1600.0  ...  2179.000000  2180.000000  2183.000000   
1466      NaN      NaN  ...  1295.000000  1295.000000  1301.000000   
1467  1555.00   1560.0  

In [99]:
# Fill in the beginning NaN's. 
for row in range(len(zillow_trim)):
    i = 7 # first rent column
    while True:
        if pd.isnull(zillow_trim.iloc[row, i]):
            i+=1
        else:
            break
    i -= 1
    while i > 6:
        zillow_trim.iloc[row, i] = (zillow_trim.iloc[row, i+12] * .9615)
        i-=1

In [100]:
zillow_trim.iloc[-50:-20, 6:]

SizeRank      2012-07      2012-08      2012-09      2012-10  \
1447      1788   570.000000   566.000000   562.000000   568.000000   
1448      1791   743.079250   631.000000   640.000000   641.000000   
1449      1792   728.000000   715.000000   703.000000   708.000000   
1450      1793   581.226750   564.400500   725.000000   711.000000   
1451      1794   691.318500   704.779500   713.000000   715.000000   
1452      1795  1181.000000  1188.000000  1195.000000  1193.000000   
1453      1796  1102.000000  1120.000000  1131.000000  1112.000000   
1454      1797   784.885430   795.054735   808.921969   825.562649   
1455      1799  2793.000000  2793.000000  2780.000000  2745.000000   
1456      1801   628.000000   633.000000   633.000000   627.000000   
1457      1802  1574.393272  1587.336023  1608.599115  1616.919455   
1458      1804  1082.667634  1085.334303  1098.667649  1112.000994   
1459      1805   868.000000   867.000000   875.000000   883.000000   
1460      1807   767.000000   760.000000   755.000000   736.000000   
1461      1808  2226.153258  2241.869456  2258.510137  2245.567385   
1462      1810  3310.570937  3354.946085  3334.607476  3332.758511   
1463      1812  2530.307918  2512.742755  2497.026557  2485.932770   
1464      1816  2067.142311  2055.124042  2061.595418  2068.066793   
1465      1818  1572.000000  1572.000000  1574.000000  1579.000000   
1466      1819  1091.813537  1105.680771  1115.850076  1115.850076   
1467      1820  1513.000000  1518.000000  1518.000000  1521.000000   
1468      1821   903.219158   906.917087   907.841569   896.747782   
1469      1822  2370.372489  2380.541794  2380.541794  2357.429738   
1470      1823   864.000000   873.000000   867.000000   870.000000   
1471      1825   747.000000   739.000000   733.000000   736.000000   
1472      1826  1732.479737  1762.987651  1781.477296  1815.683139   
1473      1827   862.541939   893.512095   924.482250   930.953626   
1474      1828  2556.193421  2592.248229  2621.831661  2631.076483   
1475      1829  1021.000000   992.000000   961.000000   930.000000   
1476      1830   890.276407   900.445712   909.690534   915.237427   

          2012-11      2012-12      2013-01      2013-02      2013-03  ...  \
1447   576.000000   577.000000   573.000000   574.000000   572.000000  ...   
1448   661.800000   682.600000   703.400000   724.200000   745.000000  ...   
1449   718.000000   728.000000   715.000000   704.000000   698.000000  ...   
1450   699.000000   680.000000   664.000000   657.000000   672.000000  ...   
1451   742.500000   770.000000   776.000000   775.000000   780.000000  ...   
1452  1191.000000  1181.000000  1170.000000  1157.000000  1158.000000  ...   
1453  1097.000000  1078.000000  1060.000000  1050.000000  1042.000000  ...   
1454   834.807472   842.203330   841.278848   810.544500   803.814000  ...   
1455  2727.000000  2749.000000  2815.000000  2882.000000  2944.000000  ...   
1456   623.000000   619.000000   624.000000   632.000000   626.000000  ...   
1457  1621.541866  1623.390831  1634.484618  1620.127500  1613.397000  ...   
1458  1122.667670  1133.334346  1139.556574  1087.191126  1107.529736  ...   
1459   883.000000   876.000000   874.000000   877.000000   879.000000  ...   
1460   719.000000   702.000000   687.000000   681.000000   689.000000  ...   
1461  2220.606365  2199.343273  2180.853628  2198.950500  2199.912000  ...   
1462  3344.776780  3370.662283  3409.490538  3439.073970  3336.405000  ...   
1463  2470.216572  2472.990019  2471.141054  2618.164500  2622.972000  ...   
1464  2083.782992  2086.556438  2086.556438  2227.795500  2210.488500  ...   
1465  1583.000000  1584.000000  1579.000000  1585.000000  1600.000000  ...   
1466  1118.623523  1108.454218  1098.284913  1138.416000  1144.185000  ...   
1467  1529.000000  1548.000000  1550.000000  1555.000000  1560.000000  ...   
1468   884.729513   896.747782   892.125371   934.578000   939.385500  ...   
1469  2356.505255  2364.825596  2362.0521

In [101]:
zillow_trim.iloc[-20:, -10:]

2019-04  2019-05  2019-06      2019-07      2019-08  2019-09  \
1477  1027.000000   1027.0   1023.0  1032.000000  1043.000000   1060.0   
1478   742.666667    719.0    705.0   697.000000   698.000000    705.0   
1479  2797.000000   2853.0   2911.0  2959.000000  2989.000000   2995.0   
1480   899.000000    894.0    897.0   905.000000   913.000000    918.0   
1481  1423.000000   1414.0   1410.0  1409.000000  1404.000000   1408.0   
1482          NaN      NaN      NaN          NaN          NaN      NaN   
1483  1111.000000   1113.0   1123.0  1130.000000  1138.000000   1141.0   
1484   871.000000    854.0    841.0   821.000000   811.000000    808.0   
1485  2592.000000   2655.0   2727.0  2787.000000  2829.000000   2850.0   
1486   807.000000    811.0    815.0   820.000000   824.000000    829.0   
1487  1922.000000   1930.0   1939.0  1946.000000  1954.000000   1958.0   
1488  2968.000000   2979.0   3001.0  3035.000000  3062.000000   3085.0   
1489  1864.000000   1866.0   1866.0  1880.000000  1894.000000   1905.0   
1490  1329.000000   1329.0   1331.0  1339.000000  1341.000000   1342.0   
1491  1092.000000   1111.0   1108.0  1105.000000  1113.000000   1113.0   
1492  2145.000000   2155.0   2166.0  2183.000000  2206.000000   2228.0   
1493   937.000000    949.0    964.0   974.000000   978.000000    978.0   
1494  2734.000000   2720.0   2745.0  2777.000000  2811.000000   2821.0   
1495   781.000000    773.0    769.0   762.000000   758.000000    758.0   
1496  1615.000000   1619.0   1628.0  1661.666667  1695.333333   1729.0   

          2019-10      2019-11  2019-12  2020-01  
1477  1068.000000  1068.000000   1045.5   1023.0  
1478   707.000000   697.000000    705.0    706.0  
1479  2922.000000  2849.000000   2832.0   2849.0  
1480   924.000000   922.000000    925.0    927.0  
1481  1434.000000  1433.000000   1423.0   1425.0  
1482          NaN          NaN      NaN      NaN  
1483  1170.000000  1157.500000   1145.0      NaN  
1484   809.000000   801.000000    793.0    806.0  
1485  2820.000000  2828.000000   2773.0   2847.0  
1486   844.000000   847.000000    850.0      NaN  
1487  1918.000000  1938.000000   1920.0   1918.0  
1488  3049.000000  3054.000000   3019.0   2992.0  
1489  1906.000000  1901.000000   1902.0   1847.0  
1490  1343.666667  1345.333333   1347.0   1350.0  
1491  1146.000000          NaN      NaN      NaN  
1492  2217.000000  2230.000000   2200.0   2192.0  
1493   998.333333  1018.666667   1039.0   1047.0  
1494  2799.000000  2950.500000   3102.0      NaN  
1495          NaN          NaN      NaN      NaN  
1496  1697.000000  1739.000000      NaN      NaN

In [102]:
# Fill in the ending NaN's. 
for row in range(len(zillow_trim)):
    i = 97 # last rent column
    while True:
        if pd.isnull(zillow_trim.iloc[row, i]):
            i-=1
        else:
            break
    i += 1
    while i < 98:
        zillow_trim.iloc[row, i] = (zillow_trim.iloc[row, i-12] * 1.04)
        i+=1

In [103]:
zillow_trim.iloc[-20:, -10:]

2019-04  2019-05  2019-06      2019-07      2019-08  2019-09  \
1477  1027.000000  1027.00  1023.00  1032.000000  1043.000000  1060.00   
1478   742.666667   719.00   705.00   697.000000   698.000000   705.00   
1479  2797.000000  2853.00  2911.00  2959.000000  2989.000000  2995.00   
1480   899.000000   894.00   897.00   905.000000   913.000000   918.00   
1481  1423.000000  1414.00  1410.00  1409.000000  1404.000000  1408.00   
1482  1457.040000  1499.68  1521.52  1526.720000  1549.600000  1571.44   
1483  1111.000000  1113.00  1123.00  1130.000000  1138.000000  1141.00   
1484   871.000000   854.00   841.00   821.000000   811.000000   808.00   
1485  2592.000000  2655.00  2727.00  2787.000000  2829.000000  2850.00   
1486   807.000000   811.00   815.00   820.000000   824.000000   829.00   
1487  1922.000000  1930.00  1939.00  1946.000000  1954.000000  1958.00   
1488  2968.000000  2979.00  3001.00  3035.000000  3062.000000  3085.00   
1489  1864.000000  1866.00  1866.00  1880.000000  1894.000000  1905.00   
1490  1329.000000  1329.00  1331.00  1339.000000  1341.000000  1342.00   
1491  1092.000000  1111.00  1108.00  1105.000000  1113.000000  1113.00   
1492  2145.000000  2155.00  2166.00  2183.000000  2206.000000  2228.00   
1493   937.000000   949.00   964.00   974.000000   978.000000   978.00   
1494  2734.000000  2720.00  2745.00  2777.000000  2811.000000  2821.00   
1495   781.000000   773.00   769.00   762.000000   758.000000   758.00   
1496  1615.000000  1619.00  1628.00  1661.666667  1695.333333  1729.00   

          2019-10      2019-11  2019-12      2020-01  
1477  1068.000000  1068.000000  1045.50  1023.000000  
1478   707.000000   697.000000   705.00   706.000000  
1479  2922.000000  2849.000000  2832.00  2849.000000  
1480   924.000000   922.000000   925.00   927.000000  
1481  1434.000000  1433.000000  1423.00  1425.000000  
1482  1588.080000  1592.240000  1602.64  1503.424000  
1483  1170.000000  1157.500000  1145.00  1154.400000  
1484   809.000000   801.000000   793.00   806.000000  
1485  2820.000000  2828.000000  2773.00  2847.000000  
1486   844.000000   847.000000   850.00   836.160000  
1487  1918.000000  1938.000000  1920.00  1918.000000  
1488  3049.000000  3054.000000  3019.00  2992.000000  
1489  1906.000000  1901.000000  1902.00  1847.000000  
1490  1343.666667  1345.333333  1347.00  1350.000000  
1491  1146.000000  1043.120000  1024.40  1056.640000  
1492  2217.000000  2230.000000  2200.00  2192.000000  
1493   998.333333  1018.666667  1039.00  1047.000000  
1494  2799.000000  2950.500000  3102.00  3020.680000  
1495   813.280000   813.280000   815.36   835.120000  
1496  1697.000000  1739.000000  1669.20  1671.626667

In [104]:
# Checking to make sure all rent NaN's are filled (there is one missing city)
zillow_trim.isnull().sum().sum()

1

In [223]:
zillow_trim

RegionID  RegionName         City State  \
0        61639       10025     New York    NY   
1        84654       60657      Chicago    IL   
2        61637       10023     New York    NY   
3        91982       77494         Katy    TX   
4        84616       60614      Chicago    IL   
...        ...         ...          ...   ...   
1540     92311       78215  San Antonio    TX   
1541     58656        2142    Cambridge    MA   
1542     65393       18232     Lansford    PA   
1543     90755       75202       Dallas    TX   
1544     84452       60301     Oak Park    IL   

                                 Metro        CountyName  SizeRank  \
0          New York-Newark-Jersey City   New York County         1   
1             Chicago-Naperville-Elgin       Cook County         2   
2          New York-Newark-Jersey City   New York County         3   
3     Houston-The Woodlands-Sugar Land     Harris County         4   
4             Chicago-Naperville-Elgin       Cook County         5   
...                                ...               ...       ...   
1540         San Antonio-New Braunfels      Bexar County      1854   
1541           Boston-Cambridge-Newton  Middlesex County      1855   
1542        Allentown-Bethlehem-Easton     Carbon County      1856   
1543       Dallas-Fort Worth-Arlington     Dallas County      1857   
1544          Chicago-Naperville-Elgin       Cook County      1859   

          2015-06      2015-07      2015-08  ...  2019-04  2019-05  2019-06  \
0     3655.000000  3685.000000  3673.000000  ...   3484.0   3523.0   3573.0   
1     1703.000000  1728.000000  1746.000000  ...   1731.0   1749.0   1768.0   
2     3508.000000  3560.000000  3596.000000  ...   3449.0   3459.0   3479.0   
3     1214.000000  1242.000000  1276.000000  ...   1226.0   1250.0   1264.0   
4     1799.000000  1808.000000  1817.000000  ...   1888.0   1911.0   1934.0   
...           ...          ...          ...  ...      ...      ...      ...   
1540  1178.714869  1208.298301  1212.920712  ...   1268.0   1301.0   1334.0   
1541  3125.000000  3137.000000  3185.000000  ...   2734.0   2720.0   2745.0   
1542   740.000000   757.000000   776.000000  ...    781.0    773.0    769.0   
1543  1457.000000  1460.000000  1467.000000  ...   1615.0   1619.0   1628.0   
1544  1350.668567  1387.647857  1392.270268  ...   1447.0   1483.0   1495.0   

          2019-07      2019-08      2019-09  2019-10  2019-11  2019-12  \
0     3622.000000  3664.000000  3698.000000  3704.00  3692.00  3715.00   
1     1787.000000  1801.000000  1806.000000  1789.00  1761.00  1747.00   
2     3491.000000  3510.000000  3521.000000  3574.00  3608.00  3561.00   
3     1251.000000  1246.000000  1252.000000  1280.00  1310.00  1294.00   
4     1950.000000  1953.000000  1952.000000  1921.00  1958.00  1937.00   
...           ...          ...          ...      ...      ...      ...   
1540  1304.000000  1267.333333  1230.666667  1194.00  1188.00  1186.00   
1541  2777.000000  2811.000000  2821.000000  2799.00  2950.50  3102.00   
1542   762.000000   758.000000   758.000000   813.28   813.28   815.36   
1543  1661.666667  1695.333333  1729.000000  1697.00  1739.00  1669.20   
1544  1511.000000  1537.000000  1570.000000  1580.00  1625.00  1476.80   

          2020-01  
0     3676.000000  
1     1731.000000  
2     3655.600000  
3     1286.000000  
4     1916.000000  
...           ...  
1540  1318.720000  
1541  3020.680000  
1542   835.120000  
1543  1671.626667  
1544  1462.240000  

[1545 rows x 63 columns]

In [105]:
zillow_trim.to_csv('Zillow_ZRI_Multifam_Trimmed_Filled_2012onwards.csv')